modules/ethereum/src/benchmarking.rs

```diff
@@ -55,6 +51,219 @@ benchmarks! {
+		let mut headers = Vec::new();
+		let mut parent = initial_header.clone();
+
+		// Import a bunch of headers without any verification, will ensure that they're not
+		// finalized prematurely
+		for i in 1..=num_blocks {
+			let header = HeaderBuilder::with_parent(&parent).sign_by(&validator(0));
+			let id = header.compute_id();
+			insert_header(&mut storage, header.clone());
+			headers.push(header.clone());
+			parent = header;
+		}
+
+		let last_header = headers.last().unwrap().clone();
+		let last_authority = validator(1);
+
+		// Need to make sure that the header we're going to import hasn't been inserted
+		// into storage already
+		let header = HeaderBuilder::with_parent(&last_header).sign_by(&last_authority);
+	}: import_unsigned_header(RawOrigin::None, header, None)
+	verify {
+		let storage = BridgeStorage::<T>::new();
+		assert_eq!(storage.best_block().0.number, (num_blocks + 1) as u64);
+		assert_eq!(storage.finalized_block().number, num_blocks as u64);
+	}
+
+	// A block import may trigger a pruning event, which adds extra work to the import progress.
+	// In this bench we trigger a pruning event in order to see how much extra time is spent by the
+	// runtime dealing with it. In the Ethereum Pallet, we're limited pruning to eight blocks in a
+	// single import, as dictated by MAX_BLOCKS_TO_PRUNE_IN_SINGLE_IMPORT.
+	import_unsigned_pruning {
+		let n in 1..100;
+		let num_headers = 10;
+
+		let mut storage = BridgeStorage::<T>::new();
+
+		let num_validators = 3;
+		let initial_header = initialize_bench::<T>(num_validators as usize);
+		let validators = validators(num_validators);
+
+		// Want to prune eligible blocks between [0, 10)
+		BlocksToPrune::put(PruningRange {
+			oldest_unpruned_block: 0,
+			oldest_block_to_keep: 10,
+		});
+
+		let mut parent = initial_header;
+		for i in 1..=num_headers {
+			let header = HeaderBuilder::with_parent(&parent).sign_by_set(&validators);
+			let id = header.compute_id();
+			insert_header(&mut storage, header.clone());
+			parent = header;
+		}
+
+		let header = HeaderBuilder::with_parent(&parent).sign_by_set(&validators);
+	}: import_unsigned_header(RawOrigin::None, header, None)
+	verify {
+		let storage = BridgeStorage::<T>::new();
+		assert_eq!(storage.best_block().0.number, (num_headers + 1) as u64);
+
+		// We're limited to pruning only 8 blocks per import
+		assert!(HeadersByNumber::get(&0).is_none());
+		assert!(HeadersByNumber::get(&7).is_none());
+	}
+
+	// The goal of this bench is to import a block which contains a transaction receipt. The receipt
+	// will contain a validator set change. Verifying the receipt root is an expensive operation to
+	// do, which is why we're interested in benchmarking it.
+	import_unsigned_with_receipts {
+		let n in 1..100;
+
+		let mut storage = BridgeStorage::<T>::new();
+
+		let num_validators = 1;
+		let initial_header = initialize_bench::<T>(num_validators as usize);
+		let receipts = vec![validators_change_receipt(Default::default())];
```

Comment:
You could add more (say `n`) receipts here to measure how receipts size affects import time.

Reply:
Ah, yeah good point!